src/Validator/Metadata/Resource/Factory/ParameterValidationResourceMetadataCollectionFactory.php

```diff
@@ -190,7 +191,18 @@ private function addSchemaValidation(Parameter $parameter, ?array $schema = null
         }
 
         if (isset($schema['type']) && 'array' === $schema['type']) {
-            $assertions[] = new Type(type: 'array');
+            $assertions[] = new Type(type: $schema['type']);
+        }
+
+        // Allow null in case of optional parameter
+        if (isset($schema['type']) && 'null' === $schema['type']) {
+            // Header null values may be sent as array through Symfony
+            $assertions[] = new Expression(expression: 'value in [null, "null", [null], ["null"]]');
+        }
+
+        // Allow null in case of optional parameter
+        if (isset($schema['type']) && 'boolean' === $schema['type']) {
+            $assertions[] = new Expression(expression: 'value in [null, 0, 1, "0", "1", false, true, "false", "true"]');
```

Comment:
I'm not a huge fan of using an expression, I'd prefer to use `Assert\IsTrue` or `Assert\IsFalse` or why not even `Assert\Type`. This leads to the same issue as above that is that parameters are always strings. Suggestion: let's add a new `castToNativeType` option on parameters, if its `true` we can cast the string to the native type inside the `ApiPlatform\State\Provider\ParameterProvider` (there's a function in type info for that IIRC). Then, if `castToNativeType` is true we can safely add an `Assert\Type` constraint.

Reply:
@soyuka does [this commit](https://github.com/api-platform/core/pull/7240/commits/f7a4639928340757d4c5fb51c15efcef3f5ec4ac) suit your proposition?